r/consensus: adjust learner initial offset to batch boundary

mmaslankaprv opened the pull request:
Redpanda Raft protocol implementation operates on batches. The batch is the smallest unit leader can send in append entries request to the followers. The fact that the batch is the smallest unit Raft operates at forces the snapshots to be taken at batch boundaries. In other words the last snapshot included offset must be a last offset of a batch.

Described invariant may not always be true when controller backend request fast reconfiguration and the offset provided is bounded by max_collectible_offset. The `max_collectible_offset` may be in the middle of the batch. (an example of state machine returning offset in the middle of the batch is datalake translation state machine which operates at the record level).

Added code that will adjust the requested learner start offset to batch boundary. The adjusted offset may not the first possible batch last offset because it uses index lower bound. (Index is sparse and using 32 KiB steps)

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
r/consensus: adjust learner initial offset to batch boundary

Redpanda Raft protocol implementation operates on batches. The batch is
the smallest unit leader can send in append entries request to the
followers. The fact that the batch is the smallest unit Raft operates at
forces the snapshots to be taken at batch boundaries. In other words the
last snapshot included offset must be a last offset of a batch.

Described invariant may not always be true when controller backend
request fast reconfiguration and the offset provided is bounded by
max_collectible_offset. The `max_collectible_offset` may be in the
middle of the batch. (an example of state machine returning offset in
the middle of the batch is datalake translation state machine which
operates at the record level).

Added code that will adjust the requested learner start offset to batch
boundary. The adjusted offset may not the first possible batch last
offset because it uses index lower bound. (Index is sparse and using
32 KiB steps)

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: added test validating initial learner offset adjustment

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/raft/consensus.cc:
```diff
@@ -1251,6 +1251,42 @@ ss::future<std::error_code> consensus::replace_configuration(
+                    _log->offsets());
+                  learner_start_offset = std::nullopt;
+              }
+              auto adjusted_learner_initial_offset = model::next_offset(
+                *maybe_adjusted_offset);
+              if (adjusted_learner_initial_offset != learner_start_offset) {
+                  vlog(
+                    _ctxlog.info,
+                    "Adjusted learner start offset {} to the batch boundary. "
+                    "Adjusted offset: {}.",
+                    learner_start_offset,
+                    adjusted_learner_initial_offset);
+              }
+              learner_start_offset = adjusted_learner_initial_offset;
```
this calculation should only be used if `maybe_adjusted_offset` has a value, right?

WillemKauf replied:
Yeah, missing `else` or some other logic here to avoid a de-reference of a `std::optional` without a value.

WillemKauf commented on src/v/raft/consensus.cc:
```diff
@@ -1251,6 +1251,42 @@ ss::future<std::error_code> consensus::replace_configuration(
+               */
+              const auto maybe_adjusted_offset = _log->index_lower_bound(
+                *learner_start_offset);
+              if (!maybe_adjusted_offset) {
```
What does this code path imply in terms of the state of the log/learner/leader?

mmaslankaprv replied:
This will effectively disable the fast reconfiguration. As the learner start offset will not be set.

WillemKauf commented on src/v/raft/consensus.cc:
```diff
@@ -1251,6 +1251,42 @@ ss::future<std::error_code> consensus::replace_configuration(
+              /**
+               * We need to adjust the offset to the full batch boundary. This
+               * is required as the last offset included in the snapshot MUST be
+               * the last offset in a batch. If that wouldn't be the case Raft
```
Am I interpreting the cover letter correctly in that this is the only location in `redpanda` in which the offset invariant described is broken (as far as we know)?

mmaslankaprv replied:
yes, exactly that is the only place. I placed the fix here to make sure the callers doesn't have to remember about the batch boundary invariant

WillemKauf commented:
Also wondering that if this was indeed an outstanding bug, what sort of issues we would run into as a result of it.

bashtanov commented:
I didn't look into `index_lower_bound` logic yet. From the comment added to `index_lower_bound` and the code added to `consensus::replace_configuration` it still looks like it still can increase `learner_start_offset`.

Imagine batches being [0..9], [10..19] and [20..29], all indexed, and original `learner_start_offset` is '{14}'. `model::prev_offset(*learner_start_offset)` is 13 then. `The maximum last offset of the indexed batch which base offset is smaller than or equal to the requested offset` would be 19, as the base offset of batch `[10..19]` is 10 and is smaller than 13. `model::next_offset(19)` is 20.

Did I misread `index_lower_bound` comment?

mmaslankaprv commented:
> Did I misread `get_term_last_offset` comment?

You are right Alexey. (the method name is `index_lower_bound`)

bashtanov commented:
Yeah sorry I copy-pasted the wrong method name :facepalm:

bharathv commented on src/v/raft/consensus.cc:
```diff
@@ -1242,6 +1242,55 @@ consensus::remove_member(vnode node, model::revision_id new_revision) {
+        return learner_start_offset;
+    }
+    /**
+     * We need to adjust the offset to the full batch boundary. This
+     * is required as the last offset included in the snapshot MUST be
+     * the last offset in a batch. If that wouldn't be the case Raft
+     * leader would not be able to recover follower as Raft delivers
+     * whole batches not individual records. We are looking for the last offset
+     * of a batch that is smaller than the requested learner_start_offset.
+     *
+     *
+     * The following invariant MUST hold:
+     * adjusted_offset <= last_included_offset
+     */
+
+    auto maybe_adjusted_offset = _log->index_lower_bound(*learner_start_offset);
+    while (maybe_adjusted_offset.has_value()
+           && maybe_adjusted_offset > learner_start_offset) {
+        maybe_adjusted_offset = _log->index_lower_bound(
+          model::prev_offset(*maybe_adjusted_offset));
+    }
```
nit: feel free to ignore

One simpler solution probably is to make a log reader with start_offset as learner_start_offset and use the begin offset of the first batch returned by the reader (hides all the complexity, I think).

mmaslankaprv replied:
you are 100% right, i was considering this but i was feeling the solution with a reader is much heavier

bashtanov replied:
Assuming the comment on `index_lower_bound` is correct, this loop may never finish. Scenario follows.

Batches [0..9]. [10..19], [20..29], all indexed.
`learner_start_offset` is {14}.
Before loop: `maybe_adjusted_offset` = {19}, as 10 <= 14 but 20 > 14.
Loop condition check: 19 > 14, pass.
Inside loop: `model::prev_offset(*maybe_adjusted_offset)` = 18;
`maybe_adjusted_offset` = {19}, as 10 <= 18 but 20 > 18.
Here we're stuck with `maybe_adjusted_offset`={19}.

mmaslankaprv replied:
I changed this part to use reader, it is not often code path and reader version is much simpler and reliable

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d4af268f-329e-4d46-b2c3-48e23fca9478-->
<!--TEST RESULTS-->
<!--d20d7df7-b9f1-4d55-84d7-cb8bbba15311-->
<!--TEST RESULTS-->
<!--1b2bd1c3-ca51-4bfa-a425-9eb12b5d81ea-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63820">build#63820</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63820#0195d94a-0731-44f1-a514-9d7f0aaba844 | FLAKY | 1/2 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=60000.topic_cfg=120000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63820#0195d94a-0731-44f1-a514-9d7f0aaba844 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63820#0195d937-1a15-4075-8769-f45debb80c8f | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63881">build#63881</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| eviction_stm_test_rpunit.eviction_stm_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63881#0195ddcc-19a9-4234-989f-22f8ca34fa4a | FAIL | 0/2 |
| eviction_stm_test_rpunit.eviction_stm_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63881#0195ddcc-19aa-4927-8139-f40d7b2da50d | FAIL | 0/2 |
| rptest.tests.archival_test.ArchivalTest.test_retention_archival_coordination.acks=-1.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.archival_test.ArchivalTest.test_retention_archival_coordination.acks=1.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.e2e_iam_role_test.AWSRoleFetchTests.test_write | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndSpilloverTest.test_spillover.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndSpilloverTest.test_spillover.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling_with_retention | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling_with_retention_consumer | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.retention_policy_test.RetentionPolicyTest.test_changing_topic_retention.property=retention.bytes.acks=1.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.retention_policy_test.RetentionPolicyTest.test_changing_topic_retention.property=retention.bytes.acks=1.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.retention_policy_test.RetentionPolicyToggleTest.test_changing_topic_cleanup_policy | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=None.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=true.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.retention_policy_test.ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention.cluster_remote_write=True.topic_remote_write=true.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.shadow_indexing_admin_api_test.SIAdminApiTest.test_bucket_validation.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f4-4c98-b079-6f490a551bc2 | FAIL | 0/20 |
| rptest.tests.shadow_indexing_firewall_test.ShadowIndexingFirewallTest.test_consume_from_blocked_s3.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
| rptest.tests.timequery_test.TimeQueryTest.test_timequery_with_spillover_gc_delayed | ducktape | https://buildkite.com/redpanda/redpanda/builds/63881#0195de0c-37f3-4647-a991-043b56b3089c | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63945">build#63945</a></summary>


| test_id                                                                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63945#0195e653-e812-4b87-bdce-84a88fdb2d77 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63945#0195e64f-c0b3-407e-a49d-5bfaa1e6247d | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63945#0195e64f-c0b1-424e-86b2-926af9a83773 | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63945#0195e653-e811-4f3f-90f6-f7c405865002 | FLAKY         | 1/2      |
</details>
<!--/body-->

ztlpn commented on src/v/raft/consensus.cc:
```diff
@@ -1242,23 +1242,84 @@ consensus::remove_member(vnode node, model::revision_id new_revision) {
+    rdr_config.skip_batch_cache = true;
+    rdr_config.max_bytes = 1;
+
+    auto rdr = co_await _log->make_reader(rdr_config);
```
Hmm, we should be able to get what we need without actually reading just by replicating the storage logic there, right?

In fact I think `disk_log_impl::index_lower_bound` implementation has a bug - usually it returns max last offset that is strictly less than the needle, but if `idx.max_offset() == o` it returns last offset that is equal (and not strictly less as in other cases).

This bug aside, I think what we need here doesn't require any prev/next offsets - just do `segments().lower_bound(o)` and then `it->idx.find_nearest(o)`. It will return what we need - max base offset that is less or equal to the needle.

ztlpn commented on src/v/raft/consensus.cc:
```diff
@@ -1251,10 +1292,13 @@ ss::future<std::error_code> consensus::replace_configuration(
           vlog(
             _ctxlog.debug,
-            "Replacing current configuration: {} with new configuration: {}",
+            "Replacing current configuration: {} with new configuration: "
+            "{}",
```
nit: unnecessary change

ztlpn commented on src/v/storage/log.h:
```diff
@@ -133,8 +133,12 @@ class log {
     virtual std::optional<model::offset>
       get_term_last_offset(model::term_id) const = 0;
+    /**
+     * Returns a base offset of the indexed batch that is smaller than or equal
```
nit: a base offset -> max base offset

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63881-->
<!--ca3f2278-d178-40c9-9c8d-cf2859365de8-->
<!--3be69d5f-f179-4297-ba3a-52952317154b-->
<!--title-->
## Retry command for [Build#63881](https://buildkite.com/redpanda/redpanda/builds/63881)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":true,"topic_remote_write":null}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":true,"topic_remote_write":"true"}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndSpilloverTest.test_spillover@{"cloud_storage_type":2}
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery_with_spillover_gc_delayed
tests/rptest/tests/archival_test.py::ArchivalTest.test_retention_archival_coordination@{"acks":1,"cloud_storage_type":1}
tests/rptest/tests/log_segment_ms_test.py::SegmentMsTest.test_segment_rolling_with_retention
tests/rptest/tests/retention_policy_test.py::RetentionPolicyTest.test_changing_topic_retention@{"acks":1,"cloud_storage_type":2,"property":"retention.bytes"}
tests/rptest/tests/retention_policy_test.py::RetentionPolicyToggleTest.test_changing_topic_cleanup_policy
tests/rptest/tests/shadow_indexing_firewall_test.py::ShadowIndexingFirewallTest.test_consume_from_blocked_s3@{"cloud_storage_type":1}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":1,"cluster_remote_write":true,"topic_remote_write":null}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":1,"cluster_remote_write":true,"topic_remote_write":"true"}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndSpilloverTest.test_spillover@{"cloud_storage_type":1}
tests/rptest/tests/e2e_iam_role_test.py::AWSRoleFetchTests.test_write
tests/rptest/tests/log_segment_ms_test.py::SegmentMsTest.test_segment_rolling_with_retention_consumer
tests/rptest/tests/archival_test.py::ArchivalTest.test_retention_archival_coordination@{"acks":-1,"cloud_storage_type":2}
tests/rptest/tests/retention_policy_test.py::RetentionPolicyTest.test_changing_topic_retention@{"acks":1,"cloud_storage_type":1,"property":"retention.bytes"}
tests/rptest/tests/shadow_indexing_admin_api_test.py::SIAdminApiTest.test_bucket_validation@{"cloud_storage_type":2}
```
<!--/body-->

mmaslankaprv committed:
log: added comment to `index_lower_bound` method

The index lower bound method is very special and used only by an
eviction state machine. Added a comment clarifying the semantics.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
log: added method allowing to find indexed batch base offset upper bound

Added `storage::log::index_batch_base_offset_lower_bound` a method
returning a `base_offset` of a batch that is smaller than or equal to
the requested offset. The method uses segment index to find the offset
so returned value may not be the first possible batch with smaller
offset but the previous one indexed.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
/backport v25.1.x
